### PR TITLE
STM32 L1 and F4 platforms merge

### PR DESCRIPTION
**Pull request by forGGe**
/cc @vadimol

### CODE REVIEW

**Comment by forGGe** [comment]
@vadimol could you please re-test current state of this PR? Especially circular mode in SPI driver

File: lib/stm32-spl/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
-if("${CONFIG_PLATFORM}" STREQUAL "stm32f4xx")
+if("${CONFIG_PLATFORM_DEVICE}" MATCHES "STM32F4")
     include(stm32f4xx_spl.cmake)
-elseif("${CONFIG_PLATFORM}" STREQUAL "stm32l1xx")
+elseif("${CONFIG_PLATFORM_DEVICE}" MATCHES "STM32L1")
     include(stm32l1xx_spl.cmake)
 endif()
```
Comment (RostakaGmfun) [inline lib/stm32-spl/CMakeLists.txt]: And what about error reporting, if unsupported `CONFIG_PLATFORM_DEVICE` was supplied?

File: platform/stm32/export/aux/i2c_bus.hpp [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+
+    //!
+    //! \brief Sets rx buffer with given size.
+    //! \param[in,out]  rx      Buffer to write data to. Optional.
+    //! \param[in]      size    Size
+    //!
+    void set_rx(uint8_t *rx, size_t size);
+
+    //!
+    //! \brief Sets rx buffer made-up from sequence of similar bytes.
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/i2c_bus.hpp]: `rx`->`tx`

File: platform/stm32/export/aux/i2c_bus.hpp [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+	,m_rx_size{0}
+	,m_rx_left{0}
+	,m_status{0}
+	,m_slave_addr{0}
+{
+
+}
+
+template< class i2c_config >
+i2c_bus< i2c_config >::~i2c_bus()
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/i2c_bus.hpp]: The body is empty. How about declaring it `default`?

File: platform/stm32/export/aux/i2c_bus.hpp [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+}
+
+template< class i2c_config >
+void i2c_bus< i2c_config >::set_handler(const handler_fn &handler)
+{
+    m_event_handler = handler;
+}
+
+template< class i2c_config >
+ecl::err i2c_bus <i2c_config>::i2c_setup_xfer_irq()
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/i2c_bus.hpp]: Spaces around angular braces seem to inconsistent

File: platform/stm32/export/aux/spi_bus.hpp [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+
+    //!
+    //! \brief Sets rx buffer with given size.
+    //! \param[in,out]  rx      Buffer to write data to. Optional.
+    //! \param[in]      size    Size
+    //!
+    void set_rx(uint8_t *rx, size_t size);
+
+    //!
+    //! \brief Sets rx buffer made-up from sequence of similar bytes.
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/spi_bus.hpp]: `rx`->`tx`

File: platform/stm32/export/aux/spi_bus.hpp [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+    ,m_tx_size{0}
+    ,m_rx{nullptr}
+    ,m_rx_size{0}
+    ,m_status{rx_complete | tx_complete}
+{
+
+}
+
+template<spi_device dev>
+spi_i2s_bus<dev>::~spi_i2s_bus()
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/spi_bus.hpp]: How about defaulting empty destructor?

File: platform/stm32/export/aux/spi_bus.hpp [hunk 7]
```diff
@@ -0,0 +1,988 @@
...
+template<spi_device dev>
+void spi_i2s_bus<dev>::set_tx(const uint8_t *tx, size_t size)
+{
+    if (!(m_status & inited)) {
+        return;
+    }
+
+    m_status &= ~(mode_fill);
+
+    // Threat nullptr buffers as indication that xfer of this type
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/spi_bus.hpp]: Treat

File: platform/stm32/export/aux/spi_bus.hpp [hunk 8]
```diff
@@ -0,0 +1,988 @@
...
+}
+
+template<spi_device dev>
+void spi_i2s_bus<dev>::set_rx(uint8_t *rx, size_t size)
+{
+    if (!(m_status & inited)) {
+        return;
+    }
+
+    // Threat nullptr buffers as indication that xfer of this type
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/spi_bus.hpp]: Treat

File: platform/stm32/export/aux/spi_bus.hpp [hunk 9]
```diff
@@ -0,0 +1,988 @@
...
+}
+
+template<spi_device dev>
+ecl::err spi_i2s_bus<dev>::do_xfer()
+{
+    if (!(m_status & inited)) {
+        return err::perm;
+    }
+
+    if (!valid_sizes()) {
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/spi_bus.hpp]: How about performing this check in `set_tx()`/`set_rx()`?

File: platform/stm32/export/aux/spi_bus.hpp [hunk 10]
```diff
@@ -0,0 +1,988 @@
...
+{
+    if (m_status & tx_hidden) {
+        // TX is not requested by user but it is required to setup this stream
+        // in SPI in order to receive data from RX channel.
+        // Trick is to use fill mode in TX channel without notifying
+        // user about it.
+
+        if (!m_rx_size) {
+            // TODO: is this correct that both buffers with 0 size???
+            for (;;);
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/spi_bus.hpp]: How about some assertion/error reporting?

File: platform/stm32/export/aux/usart_bus.hpp [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+
+    //!
+    //! \brief Sets rx buffer with given size.
+    //! \param[in,out]  rx      Buffer to write data to. Optional.
+    //! \param[in]      size    Size
+    //!
+    void set_rx(uint8_t *rx, size_t size);
+
+    //!
+    //! \brief Sets rx buffer made-up from sequence of similar bytes.
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/usart_bus.hpp]: `rx`->`tx`

File: platform/stm32/export/aux/usart_bus.hpp [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+    ,m_rx{nullptr}
+    ,m_rx_size{0}
+    ,m_rx_left{0}
+    ,m_status{0}
+{
+
+}
+
+template< usart_device dev >
+usart_bus< dev >::~usart_bus()
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/usart_bus.hpp]: Default destructor?

File: lib/stm32-spl/CMakeLists.txt (hunk 1 above)
Comment (forGGe) [inline lib/stm32-spl/CMakeLists.txt]: Will be added, thanks

File: platform/stm32/export/aux/i2c_bus.hpp (hunk 2 above)
Comment (forGGe) [inline platform/stm32/export/aux/i2c_bus.hpp]: +

File: platform/stm32/export/aux/i2c_bus.hpp (hunk 3 above)
Comment (forGGe) [inline platform/stm32/export/aux/i2c_bus.hpp]: Basic types have no default constructors, so those sub-object will not be initialized properly by default.

File: platform/stm32/export/aux/i2c_bus.hpp (hunk 4 above)
Comment (forGGe) [inline platform/stm32/export/aux/i2c_bus.hpp]: +

File: platform/stm32/export/aux/spi_bus.hpp (hunk 5 above)
Comment (forGGe) [inline platform/stm32/export/aux/spi_bus.hpp]: +

File: platform/stm32/export/aux/spi_bus.hpp (hunk 6 above)
Comment (forGGe) [inline platform/stm32/export/aux/spi_bus.hpp]: +

File: platform/stm32/export/aux/spi_bus.hpp (hunk 7 above)
Comment (forGGe) [inline platform/stm32/export/aux/spi_bus.hpp]: +

File: platform/stm32/export/aux/spi_bus.hpp (hunk 8 above)
Comment (forGGe) [inline platform/stm32/export/aux/spi_bus.hpp]: +

File: platform/stm32/export/aux/spi_bus.hpp (hunk 10 above)
Comment (forGGe) [inline platform/stm32/export/aux/spi_bus.hpp]: +_

File: platform/stm32/export/aux/usart_bus.hpp (hunk 11 above)
Comment (forGGe) [inline platform/stm32/export/aux/usart_bus.hpp]: +

File: platform/stm32/export/aux/usart_bus.hpp (hunk 12 above)
Comment (forGGe) [inline platform/stm32/export/aux/usart_bus.hpp]: +

**Pull request merged**